### PR TITLE
added godoc comments

### PR DESCRIPTION
**Pull request by zingales**
The best way to view the code is probably to run the godoc command locally and view it there. 
run `godoc -http=:6060` and go to `http://localhost:6060/pkg/github.com/atmiguel/cerealnotes/` to view what it would look like in godoc

### CODE REVIEW

File: databaseutil/databaseutil.go [hunk 1]
```diff
@@ -10,8 +16,12 @@ import (
 
 var db *sql.DB
 
+// UniqueConstraintError is returned when a uniqueness constraint is violated
+// when trying to insert into the table
 var UniqueConstraintError = errors.New("postgres: unique constraint violation")
 
+// ConnectToDatabase connects to the database and pings the database to make
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: The first two comments are useful, but saying `ConnectToDatabase connects to the database` seems a bit crazy to me

File: databaseutil/databaseutil.go [hunk 2]
```diff
@@ -10,8 +16,12 @@ import (
 
 var db *sql.DB
 
+// UniqueConstraintError is returned when a uniqueness constraint is violated
+// when trying to insert into the table
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: Missing period

File: databaseutil/databaseutil.go [hunk 3]
```diff
@@ -30,6 +40,7 @@ func ConnectToDatabase(databaseUrl string) error {
 	return nil
 }
 
+// InsertIntoUsersTable inserts the given user information into the database
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: This comment is less informative than the function def

File: databaseutil/databaseutil.go [hunk 4]
```diff
@@ -53,6 +64,8 @@ func InsertIntoUsersTable(
 	return nil
 }
 
+// GetPasswordForUserWithEmailAddress given an email address returns the password
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: Just restating the func def here

File: databaseutil/databaseutil.go [hunk 5]
```diff
@@ -71,6 +84,8 @@ func GetPasswordForUserWithEmailAddress(emailAddress string) ([]byte, error) {
 	return password, nil
 }
 
+// GetIdForUserWithEmailAddress returns the user id assosiated with the given
```
Comment (atmiguel) [inline databaseutil/databaseutil.go]: Again, restating func def

File: handlers/doc.go [hunk 6]
```diff
@@ -0,0 +1,7 @@
+/*
+Package handlers provides but authenticated and unauthenticated http handlers.
```
Comment (atmiguel) [inline handlers/doc.go]: `provides but`? typo?

File: handlers/handlers.go [hunk 7]
```diff
@@ -18,18 +18,26 @@ const oneWeek = time.Hour * 24 * 7
 const credentialTimeoutDuration = oneWeek
 const cerealNotesCookieName = "CerealNotesToken"
 
+// Type JwtTokenClaim is a struct containting the standarded jwt claims and
```
Comment (atmiguel) [inline handlers/handlers.go]: `standarded` looks like a typo

File: handlers/handlers.go [hunk 8]
```diff
@@ -18,18 +18,26 @@ const oneWeek = time.Hour * 24 * 7
...
+// any aditional claims required for authentication for these endpoints
```
Comment (atmiguel) [inline handlers/handlers.go]: Missing period

File: handlers/doc.go [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+
+All unauthenticated handlers are of type http.HandlerFunc, and all authenticated
+handlers are of type AuthenticatedRequestHandlerType
```
Comment (atmiguel) [inline handlers/doc.go]: Missing period

File: handlers/handlers.go [hunk 10]
```diff
@@ -18,18 +18,26 @@ const oneWeek = time.Hour * 24 * 7
...
 type JwtTokenClaim struct {
 	models.UserId `json:"userId"`
 	jwt.StandardClaims
 }
 
 var tokenSigningKey []byte
 
+// SetTokenSigningKey sets the global token signing key to whatever is based in.
```
Comment (atmiguel) [inline handlers/handlers.go]: `whatever is based in`?

File: handlers/handlers.go [hunk 11]
```diff
@@ -18,18 +18,26 @@ const oneWeek = time.Hour * 24 * 7
...
+// Before this method is called the signing key is the zero value of []byte
```
Comment (atmiguel) [inline handlers/handlers.go]: Does not seem like a useful comment to me

File: handlers/doc.go [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+*/
+package handlers
```
Comment (atmiguel) [inline handlers/doc.go]: why did handlers get a doc.go but databaseutil didn't?

File: handlers/handlers.go [hunk 13]
```diff
@@ -59,6 +67,8 @@ func HandleLoginOrSignupRequest(
 	}
 }
 
+// HandleUserRequest responds to POST requests by attempting to create a user
+// with the information provided in the request body
```
Comment (atmiguel) [inline handlers/handlers.go]: this comment seems unnecessary

File: handlers/handlers.go [hunk 14]
```diff
@@ -101,6 +111,8 @@ func HandleUserRequest(
 	}
 }
 
+// RedirectRequestToHome responds to all GET requests by redirecting them to the
```
Comment (atmiguel) [inline handlers/handlers.go]: what do you mean by `all GET requests`

File: handlers/handlers.go [hunk 15]
```diff
@@ -101,6 +111,8 @@ func HandleUserRequest(
...
+// home path
```
Comment (atmiguel) [inline handlers/handlers.go]: Missing period

File: handlers/handlers.go [hunk 16]
```diff
@@ -120,6 +132,11 @@ func RedirectRequestToHome(
 	}
 }
 
+// HandleSessionRequest responds to POST, and DELETE reqeusts. On POST requests
+// It attempts to validate the information in the request body. If succesfull
```
Comment (atmiguel) [inline handlers/handlers.go]: Should lowercase `It`

Comment (atmiguel) [inline handlers/handlers.go]: It `authenticates`, not `validates`. I also wouldn't say it `attempts`

Comment (atmiguel) [inline handlers/handlers.go]: `successful` typo

File: handlers/handlers.go [hunk 17]
```diff
@@ -120,6 +132,11 @@ func RedirectRequestToHome(
...
+// the response constains a cookie with a valid JWT, with information to be
```
Comment (atmiguel) [inline handlers/handlers.go]: remove comma

Comment (atmiguel) [inline handlers/handlers.go]: Actually, the stuff after `valid JWT` seems like filler to me

File: handlers/handlers.go [hunk 18]
```diff
@@ -120,6 +132,11 @@ func RedirectRequestToHome(
...
+// authenticated in future requests. On DELETE, set a cookie to expire immediatly
```
Comment (atmiguel) [inline handlers/handlers.go]: `immediately` typo

File: handlers/handlers.go [hunk 19]
```diff
@@ -120,6 +132,11 @@ func RedirectRequestToHome(
...
+// essentially deleting the cookie on the host machine
```
Comment (atmiguel) [inline handlers/handlers.go]: `host` is a strange word to use here. Perhaps `client` would work better

Comment (atmiguel) [inline handlers/handlers.go]: Missing period

File: handlers/handlers.go [hunk 20]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
 	}
 }
 
+// AuthentictedRequestHandlerType is the function signature for all authenticated
```
Comment (atmiguel) [inline handlers/handlers.go]: I believe you said this elsewhere, but I suppose repeating ourselves isn't necessarily wrong

File: handlers/handlers.go [hunk 21]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
...
+// handlers
```
Comment (atmiguel) [inline handlers/handlers.go]: Missing period

File: handlers/handlers.go [hunk 22]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
...
 type AuthentictedRequestHandlerType func(
 	http.ResponseWriter,
 	*http.Request,
 	models.UserId)
 
+// AuthenticateOrRedirectToLogin is a meta handler that resonds to all request.
```
Comment (atmiguel) [inline handlers/handlers.go]: This sentence doesn't really mean anything

File: handlers/handlers.go [hunk 23]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
...
+// It tries to authenticate the input request. If it's succesfull it call the passed
```
Comment (atmiguel) [inline handlers/handlers.go]: `input request`?

Comment (atmiguel) [inline handlers/handlers.go]: `successful` typo. `call` --> `calls`

File: handlers/handlers.go [hunk 24]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
...
+// in authenticatedHandlerFunc. On failures to authenticate it redirects to the
+// LoginOrSignup Page.
```
Comment (atmiguel) [inline handlers/handlers.go]: last time we referenced this page it was written as `login or signup page`

File: handlers/handlers.go [hunk 25]
```diff
@@ -195,11 +212,17 @@ func HandleSessionRequest(
...
+// AuthentictedRequestHandlerType is the function signature for all authenticated
+// handlers
 type AuthentictedRequestHandlerType func(
 	http.ResponseWriter,
 	*http.Request,
 	models.UserId)
 
+// AuthenticateOrRedirectToLogin is a meta handler that resonds to all request.
+// It tries to authenticate the input request. If it's succesfull it call the passed
+// in authenticatedHandlerFunc. On failures to authenticate it redirects to the
```
Comment (atmiguel) [inline handlers/handlers.go]: `On failures to authenticate` sounds weird. Perhaps `When it fails to authenticate, ...`

File: handlers/handlers.go [hunk 26]
```diff
@@ -218,6 +241,8 @@ func AuthenticateOrRedirectToLogin(
 }
 
 // Authenticated Handlers
+
+// HandleHomeRequest responds to GET requests with the home page.
```
Comment (atmiguel) [inline handlers/handlers.go]: seems a bit unecessary

File: models/doc.go [hunk 27]
```diff
@@ -0,0 +1,7 @@
+/*
+Package models provides simple models of object in the system.
+
+
```
Comment (atmiguel) [inline models/doc.go]: Why two newlines?

File: models/doc.go [hunk 28]
```diff
@@ -0,0 +1,7 @@
+/*
+Package models provides simple models of object in the system.
```
Comment (atmiguel) [inline models/doc.go]: `object` --> `objects`

File: models/doc.go [hunk 29]
```diff
@@ -0,0 +1,7 @@
...
+
+
+This package also contains useful tools to create user objects
```
Comment (atmiguel) [inline models/doc.go]: Missing period

File: models/user.go [hunk 30]
```diff
@@ -2,16 +2,22 @@ package models
 
 import "strings"
 
+// UserId is used to symbolize that this int64 is a specifically a UserId
```
Comment (atmiguel) [inline models/user.go]: this is more confusing than the line of code itself

Comment (atmiguel) [reply models/user.go]: probably doesn't need a comment

File: models/user.go [hunk 31]
```diff
@@ -2,16 +2,22 @@ package models
...
 type UserId int64
 
+// EmailAddress is a wrapper struct of the private string emailAddress to make
```
Comment (atmiguel) [inline models/user.go]: seems useful to describe it's purpose, but describing what it is seems unnecessary

File: models/user.go [hunk 32]
```diff
@@ -2,16 +2,22 @@ package models
...
+// sure that emails are always formatted properly when passed around in the system.
 type EmailAddress struct {
 	emailAddressAsString string
 }
 
+// NewEmailAddress constructs an EmailAddress objects and provides some guranteess
+// that it is a properly formated email address.
```
Comment (atmiguel) [inline models/user.go]: no need to repeat func def stuff in here. The rest is fine

File: models/user.go [hunk 33]
```diff
@@ -2,16 +2,22 @@ package models
...
 func NewEmailAddress(emailAddressAsString string) *EmailAddress {
 	return &EmailAddress{emailAddressAsString: strings.ToLower(emailAddressAsString)}
 }
 
+// String returns the email address as a string
```
Comment (atmiguel) [inline models/user.go]: no need for this

File: paths/paths.go [hunk 34]
```diff
@@ -1,4 +1,10 @@
+/*
+Package paths is a package to store all the url path constants.
```
Comment (atmiguel) [inline paths/paths.go]: `is a package` seems unnecessary

File: paths/paths.go [hunk 35]
```diff
@@ -1,4 +1,10 @@
...
+*/
 package paths
 
+// HomePath is the path for the home page url
```
Comment (atmiguel) [inline paths/paths.go]: no more informative than the name

File: paths/paths.go [hunk 36]
```diff
@@ -1,4 +1,10 @@
...
 const HomePath = "/home"
+
+// LoginOrSignupPath is the path for the login/signup page url
```
Comment (atmiguel) [inline paths/paths.go]: seems unnecessary

File: routers/routers.go [hunk 37]
```diff
@@ -1,3 +1,6 @@
+/*
+Package routers defines the path to handler pairs for the endpoint.
```
Comment (atmiguel) [inline routers/routers.go]: I would say for `each` endpoint

File: services/userservice/userservice.go [hunk 38]
```diff
@@ -1,3 +1,7 @@
+/*
+Package userservice contains functions to handle interactions of app with
```
Comment (atmiguel) [inline services/userservice/userservice.go]: `contains functions` can be removed

File: services/userservice/userservice.go [hunk 39]
```diff
@@ -9,8 +13,12 @@ import (
 	"golang.org/x/crypto/bcrypt"
 )
 
+// EmailAddressAlreadyInUseError is returned when the email that was passed in
+// cannot be used becaues it already being used by another user
```
Comment (atmiguel) [inline services/userservice/userservice.go]: seems relatively self explanatory

File: services/userservice/userservice.go [hunk 40]
```diff
@@ -9,8 +13,12 @@ import (
...
 var EmailAddressAlreadyInUseError = errors.New("Email address already in use")
 
+// StoreNewUser takes a new user information and attempts to store it into the
+// database
```
Comment (atmiguel) [inline services/userservice/userservice.go]: doesn't seem useful

File: services/userservice/userservice.go [hunk 41]
```diff
@@ -41,6 +49,8 @@ func StoreNewUser(
 	return nil
 }
 
+// AuthenticateUserCredentials validates if the email address passwrod combo
+// is valid. Returns nil on success and not nil depending on the error.
```
Comment (atmiguel) [inline services/userservice/userservice.go]: It's all in the func def

File: services/userservice/userservice.go [hunk 42]
```diff
@@ -58,6 +68,8 @@ func AuthenticateUserCredentials(emailAddress *models.EmailAddress, password str
 	return nil
 }
 
+// GetIdForUserWithEmailAddress returns the assoaited userId for a given
+// email address.
```
Comment (atmiguel) [inline services/userservice/userservice.go]: In the func def

**Review by atmiguel — CHANGES_REQUESTED** [review]
Please forgive my over-usage of the word unnecessary. Also, I think comments don't need to live by the 80 char limit rule. One line per sentence seems reasonable to me.

File: handlers/doc.go (hunk 12 above)
Comment (zingales) [reply handlers/doc.go]: I decided that if there is one file and its named exactly like the package the package documentation can go in that file. If, however, there are more than 1 file or if the file doesn't share the same name as the package it goes into doc.go

File: databaseutil/databaseutil.go (hunk 5 above)
Comment (zingales) [reply databaseutil/databaseutil.go]: If you want to write a better comment i'm all willing to change it. But I think we should document all public entities.

File: handlers/handlers.go (hunk 10 above)
Comment (zingales) [reply handlers/handlers.go]: was supposed to be whatever is passed in but that seems unnecessary.

File: handlers/handlers.go (hunk 13 above)
Comment (zingales) [reply handlers/handlers.go]: please feel free to change to non empty string that you think is appropriate.

File: handlers/handlers.go (hunk 14 above)
Comment (zingales) [reply handlers/handlers.go]: This is deleted in a future pull request so am not going to worry about it.

File: handlers/handlers.go (hunk 26 above)
Comment (zingales) [reply handlers/handlers.go]: I'm open to all non-empty string suggestions.

**Review by atmiguel — APPROVED** [review]

**Pull request merged**